pkg/controller/hvpa/hvpa_controller.go

```diff
@@ -0,0 +1,651 @@
+			if rec.ContainerName == container.Name {
+				vpaMemTarget := rec.Target.Memory().DeepCopy()
+				vpaCPUTarget := rec.Target.Cpu().DeepCopy()
+				currMem := newDeploy.Spec.Template.Spec.Containers[id].Resources.Requests.Memory().DeepCopy()
+				currCPU := newDeploy.Spec.Template.Spec.Containers[id].Resources.Requests.Cpu().DeepCopy()
+
+				log.Info("VPA", "target mem", vpaMemTarget, "target cpu", vpaCPUTarget, "vpaWeight", vpaWeight)
+
+				factor := int64(100)
+				scale := int64(float64(vpaWeight) * float64(factor))
+
+				//origMem := currMem.DeepCopy()
+				vpaMemTarget.Sub(currMem)
+				diffMem := resource.NewQuantity(vpaMemTarget.Value()*scale/factor, vpaMemTarget.Format)
+				currMem.Add(*diffMem)
+				weightedMem := currMem
+				minDeltaMem, _ := resource.ParseQuantity("500M")
+
+				//origCPU := currCPU.DeepCopy()
+				vpaCPUTarget.Sub(currCPU)
+				diffCPU := resource.NewQuantity(vpaCPUTarget.ScaledValue(-3)*scale/factor, vpaCPUTarget.Format)
+				diffCPU.SetScaled(diffCPU.Value(), -3)
+				currCPU.Add(*diffCPU)
+				weightedCPU := currCPU
+				minDeltaCPU, _ := resource.ParseQuantity("300m")
+
+				log.Info("VPA", "weighted target mem", weightedMem, "weighted target cpu", weightedCPU)
+				//if weightedMem.Cmp(origMem) > 0 {
+				if diffMem.Sign() > 0 && diffMem.Cmp(minDeltaMem) > 0 {
+					// If the difference is greater than minimum delta
+					newDeploy.Spec.Template.Spec.Containers[id].Resources.Requests[corev1.ResourceMemory] = weightedMem.DeepCopy()
+					resourceChange = true
+				}
+				//if weightedCPU.Cmp(origCPU) > 0 {
+				if diffCPU.Sign() > 0 && diffCPU.Cmp(minDeltaCPU) > 0 {
+					// If the difference is greater than minimum delta
+					newDeploy.Spec.Template.Spec.Containers[id].Resources.Requests[corev1.ResourceCPU] = weightedCPU.DeepCopy()
+					resourceChange = true
+				}
+				// TODO: Add conditions for other resources also: ResourceStorage, ResourceEphemeralStorage,
+				break
+			}
+		}
+	}
+	log.Info("VPA", "vpa made changes?", resourceChange)
+	if resourceChange {
+		return true, r.Update(context.TODO(), newDeploy)
+	}
+	return false, nil
+}
+
+func (r *ReconcileHvpa) scaleIfRequired(hpaStatus *autoscaling.HorizontalPodAutoscalerStatus, vpaStatus *vpa_api.VerticalPodAutoscalerStatus, hvpa *autoscalingv1alpha1.Hvpa, deployment *appsv1.Deployment) (int32, bool, autoscalingv1alpha1.VpaWeight, error) {
+
+	currentReplicas := *deployment.Spec.Replicas
+	var desiredReplicas int32
+	if hpaStatus == nil {
+		desiredReplicas = currentReplicas
+	} else {
+		desiredReplicas = hpaStatus.DesiredReplicas
+	}
+
+	var vpaWeight autoscalingv1alpha1.VpaWeight
+	// lastFraction is set to default 1 to handle the case when vpaWeight is 1 in the matching interval,
+	// and there are no fractional vpaWeights in the previous intervals. So we need to default to this value
+	lastFraction := autoscalingv1alpha1.VpaWeight(1)
+	lookupNextFraction := false
+
+	for _, interval := range hvpa.Spec.WeightBasedScalingIntervals {
```

Comment:
If this for-loop is to pick the right interval/bucket it might be better to wrap this in a function.

Reply:
done